### ISSUE TITLE
Sort out what's happening with innerHTML on <script> and <style>

### ISSUE THREAD

**Issue by bzbarsky** [w3c/DOM-Parsing#6]
In at least Chrome, Safari, and Firefox, it looks like this:

    var style = document.createElementNS("http://www.w3.org/1999/xhtml", "style");
    style.innerHTML = "<foo></foo>";

will create a textnode with the text `"<foo></foo>"` as a child of the `<style>` element.  Even if this is happening in an XML/XHTML document.

Per spec, I would think in an XML document this should create a child element named "foo", no?

I haven't had a chance to check what IE/Edge do here yet.  /cc @annevk @smaug----

**Comment by travisleithead** [comment]
In Edge, with the following saved as an XML file:
```html
<root>
  <script xmlns="http://www.w3.org/1999/xhtml" type="text/javascript"><![CDATA[
    onload = function() {
      var style = document.createElementNS("http://www.w3.org/1999/xhtml", "style");
      style.innerHTML = "<foo></foo>";
      document.documentElement.appendChild(style);
      alert(style.firstChild);
    }
    ]]>
  </script>
</root>
```
I get an `HTMLUnknownElement`. Confirmed that in Chrome, Firefox, I get a `#text` node...

**Comment by bzbarsky** [comment]
@travisleithead, thank you for checking that!  If Edge is following the spec here and hasn't run into compat issues maybe we should just align Chrome and Firefox with the spec too instead of changing the spec.

**Comment by rniwa** [comment]
This behavior comes from https://bugs.webkit.org/show_bug.cgi?id=32962 in WebKit so there's definitely a compatibility requirement for the HTML parser but perhaps it's not the case for the XML parser.

**Comment by bzbarsky** [comment]
For the HTML parser all the engines agree and so does the spec.  This issue is just about the XML case.

**Comment by rniwa** [comment]
I know, and I think we can probably make this change although having a different behaviors in XHTML and XML would be rather confusing for authors as well so I'm not certain if this change is desirable.  It seems like Gecko, Blink, and WebKit all behaved this way for years so I wouldn't be surprised if there was a compatibility issue we aren't aware.

**Comment by bzbarsky** [comment]
I should note that the behavior in Blink and WebKit is totally broken because while the _setter_ acts just like HTML the _getter_ does not.  In Gecko, the getter does as well.  I filed a Blink bug at https://bugs.chromium.org/p/chromium/issues/detail?id=605973 and a WebKit bug at https://bugs.webkit.org/show_bug.cgi?id=157052

It's _possible_ that there is a compat issue if we change this, but I rather doubt it given that Edge doesn't have the behavior in question.

**Comment by bzbarsky** [comment]
Oh, and for the XML case... if you take the string and prepend `"<script>"` and append `"</script>"` and set that as innerHTML on some element, then of course the string is parsed differently from setting it directly as innerHTML on a `<script>`, in Gecko/Blink/WebKit.  _That_ seems confusing for authors to me!

**Comment by ArkadiuszMichalski** [comment]
@bzbarsky You mean something like this:
```js
<script>
	var str = "<div xmlns='test'></div><p></p>";
	var doc = document.implementation.createDocument("http://www.w3.org/1999/xhtml", "body", null);
	var doc2 = document.implementation.createDocument("", "body", null);

	doc.documentElement.innerHTML = "<script>" + str +"<\/script>";
	console.log(doc.documentElement.firstChild.childNodes.length);
	doc.documentElement.firstChild.innerHTML = str;
	console.log(doc.documentElement.firstChild.childNodes.length);
	
	doc2.documentElement.innerHTML = "<script>" + str +"<\/script>";
	console.log(doc2.documentElement.firstChild.childNodes.length);
	doc2.documentElement.firstChild.innerHTML = str;
	console.log(doc2.documentElement.firstChild.childNodes.length);
</script>
```
And results:
Gecko: `2 1 2 2`
Blink: `2 1 2 1`
Edge/IE11/Presto: `2 2 throw`

**Comment by bzbarsky** [comment]
Yes, like that.  But you should explicitly create the `<script>` with the namespace you want instead of assuming something about what createDocument will do, because createDocument is totally not interoperable in terms of namespaces, as you can see from comparing the Gecko and Blink results.

**Comment by ArkadiuszMichalski** [comment]
It's strange, `body` has `"http://www.w3.org/1999/xhtml"` namespace so parsing some content for it should not use that namespace? But even when I use just `<script>` Gecko/IE11/Edge/Presto create this script with ``"http://www.w3.org/1999/xhtml"`` namespace (I update above test to show namespace for first child of document element).

Ok, so when change to `<script xmlns='http://www.w3.org/1999/xhtml'>` engines will product:
Gecko: `2 1 http://www.w3.org/1999/xhtml 2 1 http://www.w3.org/1999/xhtml`
Blink: `2 1 http://www.w3.org/1999/xhtml 2 1 http://www.w3.org/1999/xhtml`
Edge/IE11/Presto: `2 2 http://www.w3.org/1999/xhtml throw`

**Comment by bzbarsky** [comment]
Right.  And my point is that the output should be "2 2", not "2 1".

**Comment by bzbarsky** [comment]
Or rather should be "2 2" per current spec, but clearly isn't in implementations.

**Comment by ArkadiuszMichalski** [comment]
From xml perspective `2 2` would be more consistent, but if  '2 1' will be documented and implemented then why not... in this specification and current implementations there are more inaccuracies.

**Comment by bzbarsky** [comment]
@hsivonen do you want to take a shot at what the spec should say here to match implementation reality?

**Comment by hsivonen** [comment]
OK. I'll take a look.
